pkg/daemon/vm_controller.go

```diff
@@ -75,80 +90,230 @@ func (r *VMReconciler) reconcile(ctx context.Context, vm *virtv1alpha1.VirtualMa
+				}
+			case virtv1alpha1.VirtualMachineMigrationRunning:
+				if vm.Status.NodeName == r.NodeName {
+					var vmPod corev1.Pod
+					vmPodKey := types.NamespacedName{
+						Name:      vm.Status.VMPodName,
+						Namespace: vm.Namespace,
+					}
+					if err := r.Get(ctx, vmPodKey, &vmPod); err != nil {
+						return fmt.Errorf("get VM Pod: %s", err)
+					}
+
+					if vmPod.Status.Phase == corev1.PodSucceeded {
+						vm.Status.Migration.Phase = virtv1alpha1.VirtualMachineMigrationSent
+					} else if migrationControlBlock.SendDomainErrCh == nil {
+						vm.Status.Migration.Phase = virtv1alpha1.VirtualMachineMigrationFailed
 					} else {
-						r.Recorder.Eventf(vm, corev1.EventTypeNormal, "Resumed", "Resumed VM")
+						select {
+						case err := <-migrationControlBlock.SendDomainErrCh:
+							if err != nil {
+								r.Recorder.Eventf(vm, corev1.EventTypeWarning, "FailedMigrate", "Failed to migrate VM to %s: %s", vm.Status.Migration.TargetNodeName, err)
+								vm.Status.Migration.Phase = virtv1alpha1.VirtualMachineMigrationFailed
+							}
+						default:
+							return errors.New("VM is migrating")
+						}
+					}
+					if sendDomainCancelFunc := migrationControlBlock.SendDomainCancelFunc; sendDomainCancelFunc != nil {
+						sendDomainCancelFunc()
 					}
-				default:
-					// ignored
 				}
+			case virtv1alpha1.VirtualMachineMigrationSent:
+				if vm.Status.Migration.TargetNodeName == r.NodeName {
+					// TODO handle VmNotCreated error
+					vmInfo, err := r.getMigrationTargetCloudHypervisorClient(vm).VmInfo(ctx)
+					if err != nil {
+						return err
+					}
+					switch vmInfo.State {
+					case "Running":
+						vm.Status.Migration.Phase = virtv1alpha1.VirtualMachineMigrationSucceeded
+						vm.Status.NodeName = vm.Status.Migration.TargetNodeName
+						vm.Status.VMPodName = vm.Status.Migration.TargetVMPodName
+						vm.Status.VMPodUID = vm.Status.Migration.TargetVMPodUID
+					default:
+						return errors.New("VM is migrating")
+					}
 
-				vm.Status.PowerAction = ""
+					if receiveDomainCancelFunc := migrationControlBlock.ReceiveDomainCancelFunc; receiveDomainCancelFunc != nil {
+						receiveDomainCancelFunc()
+					}
+				}
 			}
-		} else {
-			vm.Status.Phase = virtv1alpha1.VirtualMachineSucceeded
+			r.migrationControlBlocks[vm.UID] = migrationControlBlock
+			// TODO cleanup migrationControlBlock
 		}
-	default:
-		// ignored
 	}
 	return nil
 }
 
 func (r *VMReconciler) getCloudHypervisorClient(vm *virtv1alpha1.VirtualMachine) *cloudhypervisor.Client {
-	return cloudhypervisor.NewClient(fmt.Sprintf("/var/lib/kubelet/pods/%s/volumes/kubernetes.io~empty-dir/virtink/ch.sock", string(vm.Status.VMPodUID)))
+	return cloudhypervisor.NewClient(filepath.Join(getVMSocketDirPath(vm), "ch.sock"))
+}
+
+func getVMSocketDirPath(vm *virtv1alpha1.VirtualMachine) string {
+	return filepath.Join("/pods", string(vm.Status.VMPodUID), "volumes/kubernetes.io~empty-dir/virtink/")
```

Comment:
Keep it the old path?

Reply:
because there is a limit of unix domain socket path length, change it to `/pods` is for make it shorter. see more: https://unix.stackexchange.com/questions/367008/why-is-socket-path-length-limited-to-a-hundred-chars

Reply:
I'm aware this limitation, but the old path was still in the range.

Reply:
the prefix, such as `/var/lib/kubelet/pods/61f38851-f9e1-49ec-9f4a-2bae1c377d67/volumes/kubernetes.io~empty-dir/virtink/` has 99 chars, if the limitation is 104, the socket file only have 5 chars, in some OS `ch.sock` is invalid too. and for `receive-migration.sock` and `send-migration.sock` are too long.

Reply:
Keeping it no longer than 107 is safe. `receive-migration.sock` and `send-migration.sock` is way too long, let's make it `rx.sock` and `tx.sock` ?